CONTRIBUTING.md

```diff
@@ -0,0 +1,135 @@
+
+Developers must follow these steps to make a change:
+
+1. Fork the `openshift/ocs-operator` repository on GitHub.
+2. Create a branch from the master branch, or from a versioned branch (such
+   as `release-4.2`) if you are proposing a backport.
+3. Make changes.
+4. Create tests as needed and ensure that all tests pass.
+5. Ensure your commit messages include sign-off.
+6. Push your changes to a branch in your fork of the repository.
+7. Submit a pull request to the `openshift/ocs-operator` repository.
+8. Work with the community to make any necessary changes through the code
+   review process (effectively repeating steps 3-8 as needed).
+
+## Developer Environment Installation
+
+Instructions to create a dev environment for OCS-operator can be found in the
+[main project documentation](./README.md#installation-of-development-builds).
+
+## Commit Requirements
+
+In addition to passing [functional tests](./README.md#functional-tests), all
+commits to OCS-operator must follow the guidelines in this section. You can
+verify your changes meet most of these standards by running the following
+command in the repository's root directory:
+
+```
+make ocs-operator-ci
+```
+
+### Commits Per Pull Request
+
+*This requirement cannot be tested by make ocs-operator-ci.*
+
+OCS-operator is a project which maintains several versioned branches
+independently. When backports are necessary, monolithic commits make it
+difficult for maintainers to cleanly backport only the necessary changes.
+
+Pull requests should always represent a complete logical change. Where
+possible, though, pull requests should be composed of multiple commits that
+each make small but meaningful changes. Striking a balance between minimal
+commits and logically complete changes is an art as much as a science, but
+when it is possible and reasonable, divide your pull request into more commits.
+
+Some times when it will almost always make sense to separate parts of a change
+into their own commits are:
+- Changes to unrelated formatting and typo-fixing.
+- Refactoring changes that prepare the codebase for your logical change.
+- Similar changes to multiple parts of the codebase (for instance, similar
+  changes to handling of CephFilesystem, CephBlockPool, and CephObjectStore).
+
+Even when breaking down commits, each commit should leave the codebase in a
+working state. The code should add necessary unit tests and pass unit tests,
+formatting tests, and usually functional tests. There can be times when
+exceptions to these requirements are appropriate (for instance, it is sometimes
+useful for maintainability to split code changes and related changes to CRDs
+and CSVs). Unless you are very sure this is true for your change, though, make
+sure each commit passes CI checks as above.
+
+### Commit and Pull Request Messages
+
+*This requirement cannot be fully tested by make ocs-operator-ci.*
+
+All commit messages must contain:
+- A single summary line describing the change
```

Comment:
Would it be a good idea to remind developers that the summary line should be within a character limit? It need not be a strict number but I have seen commits with summary line > 150 chars and we should discourage that.

Reply:
Yes! It should be <= 80 characters, afaik. The same holds for the wrapping of the paragraphs of the body.
Many projects even use <= 72 chars.

I would even go further and prescribe a bit of grammar etc:

- Use the imperative form for the subject line. Don't use the present tense or past tense.
   I.e. use "Fix bug foobar" instead of "Fixed but foobar" or "Fixes bug foobar".

   This matches commit messages generated by `git merge` and `git revert` etc...
   A properly formed git commit subject line should always be able to complete the sentence
  `If applied, this commit will <your subject line here>`

- capitalize the subject line (possibly after the colon following the section)

- don't end the subject line with a period

Reply:
@egafford @jarrpa - this is BTW one of the thing that drives me up the wall most, the use of different tenses in the subject.
Can we agree on and write it down in the guide to require the "imperative" form as described above? :smiley:

Reply:
> @egafford @jarrpa - this is BTW one of the things that drive me up the wall most: the use of different tenses in the subject.
> Can we agree on and write it down in the guide to require the "imperative" form as described above? smiley
>
That'll affect me the most :cry: but I guess that makes sense.

I was following `this commit adds/updates <your changes here>` format which I picked up from some other project.

Reply:
Huh, neat! I didn't actually know about the imperative mood preference in Git commit messages. Good to know; I'll add that, Also yes, I've most often seen <= 80 chars; if we're comfortable with that, I'll go ahead and add it in.

Reply:
While I completely agree with your grammatical and syntactic requirements, I am on the fence about making it a hard requirement. I would take them as recommendations, but I wouldn't want to reject a PR for having commits that break these conventions, especially if nothing else is wrong.

Reply:
This grammar aspect is anyway the hardest bit to enforce by software.
Furthermore, GitHub makes it notoriously hard to review commit messages. It does not seem to be a preference here.

For me, Umanga's approach is the second most reasonable.
But I would mainly propose uniformity. 

And @umangachapagain is not the only or most affected contributor. @anmolsachan, @kesavanvt, and @rohan47   have been using past tense. @aruniiird use the gerund (?) - ing-form.

But apart from me,  @jarrpa, @raghavendra-talur, @iamniting, @rajatsing, @crombus , @nb-ohad, @nbalacha, @dannyzaken, @agarwal-mudit , @rohantmp , @travisn, etc have all been using the imperative form. (There are certainly more missing.)

And apart from this, the argument that git's own commits (Merge/Revert) use the imperative form, as well as most projects I have worked with, decides it for me.

I am fine with stating it as a recommendation **for a start**. We can still make it more strict later.
And I would certainly be pointing folks at the doc in reviews for this purpose too.